File: hedera-mirror-rest/__tests__/stream/utils.test.js
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2021 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

'use strict';

const long = require('long');
const {proto} = require('@hashgraph/proto/lib/proto');
const utils = require('../../stream/utils');

describe('protoTransactionIdToString', () => {
  const accountID = {
    shardNum: 0,
    realmNum: 0,
    accountNum: 1010,
  };

  const testSpecs = [
    {
      transactionId: proto.TransactionID.create({
        accountID,
        transactionValidStart: {
          seconds: 193823,
          nanos: 0,
        },
      }),
      expected: '0.0.1010-193823-0',
    },
    {
      transactionId: proto.TransactionID.create({
        accountID,
        transactionValidStart: {
          seconds: 193823,
          nanos: 999999999,
        },
      }),
      expected: '0.0.1010-193823-999999999',
    },
    {
      transactionId: proto.TransactionID.create({
        accountID,
        transactionValidStart: {
          seconds: long.MAX_VALUE,
          nanos: 999999999,
        },
      }),
      expected: `0.0.1010-${long.MAX_VALUE.toString()}-999999999`,
    },
  ];

  testSpecs.forEach((testSpec) => {
    test(`expect output ${testSpec.expected}`, () => {
      expect(utils.protoTransactionIdToString(testSpec.transactionId)).toEqual(testSpec.expected);
    });
  });
});

describe('readLengthAndBytes', () => {
  const data = new Array(16).fill(0xab);
  const bufferWithChecksum = Buffer.from([0, 0, 0, 16, 0, 0, 0, 101 - 16].concat(data));
  const bufferNoChecksum = Buffer.from([0, 0, 0, 16].concat(data));

  const testSpecs = [
    {
      name: 'read from buffer with checksum',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: 1,
      maxLength: data.length,
      expected: {
        length: bufferWithChecksum.length,
        bytes: Buffer.from(data),
      },
    },
    {
      name: 'read from buffer with checksum and exact length',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: data.length,
      maxLength: data.length,
      expected: {
        length: bufferWithChecksum.length,
        bytes: Buffer.from(data),
      },
    },
    {
      name: 'read from buffer without checksum',
      buffer: bufferNoChecksum,
      hasChecksum: false,
      minLength: 1,
      maxLength: data.length,
      expected: {
        length: bufferNoChecksum.length,
        bytes: Buffer.from(data),
      },
    },
    {
      name: 'read from buffer with checksum error',
      buffer: bufferNoChecksum,
      hasChecksum: true,
      minLength: 1,
      maxLength: data.length,
    },
    {
      name: 'read from buffer with length over max',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: 1,
      maxLength: data.length - 1,
    },
    {
      name: 'read from buffer with length smaller than min',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: data.length + 1,
      maxLength: data.length + 2,
    },
    {
      name: 'read from buffer with length not match',
      buffer: bufferWithChecksum,
      hasChecksum: true,
      minLength: data.length + 1,
      maxLength: data.length + 1,
    },
  ];

  testSpecs.forEach((testSpec) => {
    const {name, buffer, minLength, maxLength, hasChecksum, expected} = testSpec;
    test(name, () => {
      if (expected) {
        const actual = utils.readLengthAndBytes(buffer, minLength, maxLength, hasChecksum);
        expect(actual).toEqual(expected);
      } else {
        expect(() =>
          utils.readLengthAndBytes(buffer, minLength, maxLength, hasChecksum)
        ).toThrowErrorMatchingSnapshot();
      }
    });
  });
});

describe('readNBytes', () => {
  const buffer = Buffer.from(new Array(16).fill(0xab));
  const testSpecs = [
    {
      name: 'read 1 byte from 16-byte buffer',
      buffer,
      length: 1,
      expected: buffer.slice(0, 1),
    },
    {
      name: 'read 16 byte from 16-byte buffer',
      buffer,
      length: 16,
      expected: buffer,
    },
    {
      name: 'read 0 byte from 16-byte buffer',
      buffer,
      length: 0,
      expected: buffer.slice(0, 0),
    },
    {
      name: 'read 17 byte from 16-byte buffer',
      buffer,
      length: 17,
    },
    {
      name: 'read -1 byte from 16-byte buffer',
      buffer,
      length: -1,
    },
  ];

  testSpecs.forEach((testSpec) => {
    test(testSpec.name, () => {
      if (testSpec.expected) {
        expect(utils.readNBytes(testSpec.buffer, testSpec.length)).toEqual(testSpec.expected);
      } else {
        expect(() => utils.readNBytes(testSpec.buffer, testSpec.length)).toThrowErrorMatchingSnapshot();
      }
    });
  });
});

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/state/UtilsTest.java
/*
 * Copyright (C) 2024-2025 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.web3.state;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.hedera.hapi.node.base.Key;
import com.hedera.hapi.node.base.Timestamp;
import com.hedera.mirror.common.domain.DomainBuilder;
import com.hedera.mirror.common.util.DomainUtils;
import com.hederahashgraph.api.proto.java.Key.KeyCase;
import java.time.Instant;
import org.hyperledger.besu.datatypes.Address;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class UtilsTest {

    @Test
    void parseEd25519PublicKey() {
        final var domainBuilder = new DomainBuilder();
        final var rawKey = domainBuilder.key(KeyCase.ED25519);
        final var parsedKey = Utils.parseKey(rawKey);

        assertThat(parsedKey).isNotNull();
        assertThat(parsedKey.hasEd25519()).isTrue();
    }

    @Test
    void parseEcdsaSecp256k1PublicKey() {
        final var domainBuilder = new DomainBuilder();
        final var rawKey = domainBuilder.key(KeyCase.ECDSA_SECP256K1);
        final var parsedKey = Utils.parseKey(rawKey);

        assertThat(parsedKey).isNotNull();
        assertThat(parsedKey.hasEcdsaSecp256k1()).isTrue();
    }

    @Test
    void parseInvalidKey() {
        byte[] invalidKeyBytes = new byte[] {};

        Key parsedKey = Utils.parseKey(invalidKeyBytes);

        assertNull(parsedKey);
    }

    @Test
    void parseThrowsExceptionAndReturnsNull() {
        byte[] invalidKeyBytes = new byte[] {0x01, 0x02, 0x03, 0x04};

        Key parsedKey = Utils.parseKey(invalidKeyBytes);

        assertNull(parsedKey);
    }

    @Test
    void testConvertToTimestamp() {
        // Given
        Instant expectedInstant = Instant.now();

        long expectedEpochSecond = expectedInstant.getEpochSecond();
        int expectedNano = expectedInstant.getNano();
        long timestampNanos = DomainUtils.convertToNanos(expectedEpochSecond, expectedNano);

        // When
        Timestamp result = Utils.convertToTimestamp(timestampNanos);

        // Then
        assertThat(result.seconds()).isEqualTo(expectedEpochSecond);
        assertThat(result.nanos()).isEqualTo(expectedNano);
    }

    @Test
    void isMirrorAddressReturnsTrue() {
        final var address = Address.fromHexString("0x00000000000000000000000000000000000004e4");
        assertTrue(Utils.isMirror(address));
    }

    @Test
    void isMirrorByteArrayAddressReturnsTrue() {
        final var byteArrayAddress = new byte[] {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, -28};
        assertTrue(Utils.isMirror(byteArrayAddress));
    }

    @Test
    void isMirrorByteArrayAddressDoesNotStartWithZeroesReturnsFalse() {
        final var byteArrayAddress = new byte[] {1, 2, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, -28};
        assertFalse(Utils.isMirror(byteArrayAddress));
    }

    @Test
    void isMirrorAddressLengthIncorrectReturnsFalse() {
        byte[] address = new byte[] {0x01};
        assertFalse(Utils.isMirror(address));
    }
}
